Import users

mandryllo opened the pull request.

Commit:
Created import dialog

Commit:
Merge branch 'develop' into feature/import-users

Commit:
Fixes on import dialog

Commit:
Returning import erros and cleanup

Commit:
Refactored backend

- multer now uses memory storage
- exceljs is used for reading files
- create user uses invite function

Commit:
Import errors are returned as .xlsx file

Commit:
Backend cleanup

Commit:
Added importing progress on import dialog

vladimyr commented on client/admin/api/user.js:
```diff
@@ -23,9 +24,19 @@ function remove(item) {
+    responseType: 'arraybuffer',
+    url: url.import,
+    data: items
+  });
```
Use `request.post(url, config)` form :bowtie: 
You are expecting file as response? If so use `responseType: blob` instead 😉

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+            v-model="filename"
+            :error-messages="showErrors"
+            :disabled="importing"
+            @click="$refs.fileInput.click()"
```
Don't use javascript for this. Instead wrap both `<v-text-field>` and `<input type="file">` with `<label>` element. Click on label will activate inner input :bowtie:

Commit:
:lipstick:

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+            ref="fileInput"
+            @change="onFileSelected"
+            data-vv-name="file"
+            type="file"/>
```
Don't _selfclose_ html tags. It's not required and makes reader think that they are looking at vue component instead :bowtie:

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+          <v-btn
+            :disabled="disabled"
+            color="success"
+            type="submit">
```
This fits under 80 columns 😉

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+      filename: null,
+      form: null,
+      errors: null,
+      rules: {
```
Rules are constant so extract them to `const rules = ...` and just inject here.

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+    }
+  },
+  methods: {
+    onFileSelected($event) {
```
Use `e` instead of `$event`. It isn't injected, it is simply passed 😉

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+    onFileSelected($event) {
+      this.form = new FormData();
+      this.errors = null;
+      const file = head([...$event.target.files]);
```
Why did you do this? 🤔

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,171 @@
+      this.$validator.validateAll().then(isValid => {
+        if (!isValid || this.errors) return;
+        this.importing = true;
+        api.bulkImport(this.form).then(response => {
```
Return here and chain the rest onto existing promise chain.

vladimyr commented on server/user/import.js:
```diff
@@ -0,0 +1,37 @@
+'use strict';
+
+const Excel = require('exceljs');
```
Destructure `Workbook`? :bowtie:

Commit:
Merge branch 'develop' into feature/import-users

vladimyr commented on server/user/import.js:
```diff
@@ -0,0 +1,37 @@
+  const headers = [];
+  const data = [];
+  sheet.eachRow((row, rowIndex) => {
+    if (rowIndex === 1) return row.eachCell(cell => headers.push(cell.value));
```
Reading through [source code](https://github.com/guyonroche/exceljs/blob/db136ab0c53ce2afeb39d3359b9aaa13ecd225fe/lib/doc/row.js#L154) I've found easier way:
```js
const keys = sheet.getRow(0).values
```

vladimyr commented on server/user/import.js:
```diff
@@ -0,0 +1,37 @@
+    const properties = [];
+    const it = {};
+    row.eachCell((cell, cellIndex) => properties.push(cell.value));
+    properties.forEach((prop, index) => (it[headers[index]] = prop));
```
You can further simplify it:
```js
return row.values.reduce((acc, value, i) => {
  return Object.assign(acc, { keys[i]: value });
}, {});
```

vladimyr commented on server/user/import.js:
```diff
@@ -0,0 +1,37 @@
+function sheetToJson(sheet) {
+  const headers = [];
+  const data = [];
+  sheet.eachRow((row, rowIndex) => {
```
```js
const rowValues = sheet.getSheetValues().slice(1);
return rowValues(values => ... );
```

Commit:
Import Dialog component cleanup

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,168 @@
+    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+    'application/vnd.ms-excel',
+    'text/csv'
+  ]
```
[Check point No 3](https://blog.codinghorror.com/new-programming-jargon/) and apply it here on array brackets :wink:

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,168 @@
+      this.form.append('file', file, file.name);
+      this.$validator.validateAll().then(isValid => {
+        if (!isValid) return (this.disabled = true);
+        this.disabled = false;
```
Well, this is actually same as:
```js
.then(isValid => (this.disabled = !isValid))
```

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,168 @@
+      this.$validator.validateAll().then(isValid => {
+        if (!isValid || this.errors) return;
+        this.importing = true;
+        return api.bulkImport(this.form).then(response => {
```
Don't nest promise chains, instead append onto existing one.

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -0,0 +1,168 @@
+    },
+    downloadErrorsFile() {
+      JSZip.loadAsync(this.errors)
+        .then(zip => zip.generateAsync({ type: 'blob' }))
```
Help me understand this one :thinking: `JSZip` is used here to load zip blob? and return it without modifying it? :confused:

vladimyr replied:
Btw couldn't you just return error file as part of server response thus avoiding increasing clientside bundle by `JSZip` inclusion? :thinking:

vladimyr commented on server/user/user.controller.js:
```diff
@@ -87,12 +90,39 @@ function resetPassword({ body, params }, res) {
+    return User.invite(pick(user, inputAttrs), { existingUser, origin })
+      .catch(err => errors.push(`User ${user.email}: ${err.message}!`));
+  });
+  if (errors.length) {
```
Swap condition and make early return to avoid indentation :wink:

Commit:
Removed JSZip package

Commit:
Wrapped input elements inside label

Commit:
Added TableSheet class and refactored backend

Commit:
Reading from sheet moved to TableSheet

underscope committed:
Cleanup :lipstick:

vladimyr committed:
[WIP] Refactoring user import mechanism

- implementing generic Datasheet class capable of loading/writing
  datasets from/to tabular workbooks (csv or xlsx)
- importing logic extracted to separate module, attached to User model
  (User#import)
- calculating report filename instead of hardcoding it on client side
  :bug:
- making sure that report format matches input workbook format

vladimyr committed:
Cleanup :lipstick:

vladimyr committed:
Refactor :wrench:

- inlining importing logic into User model

vladimyr committed:
Enable column filtering

- implementing column filtering mechanism for loading data from
  Datasheets

vladimyr committed:
Improve input/output format hadling

- removing xls mimetype from list of supported types (import dialog)
- allowing explicit format override for import errors report
- reusing email validation message for registration error

vladimyr committed:
Skip header row while importing :bug:

Commit:
Merge branch 'develop' into feature/import-users

Commit:
Changed icons to mdi webfont

Commit:
Cleanup :lipstick:

Commit:
Refactored frontend

vladimyr committed:
Optimize user import :rocket:

- minimizing number of executed queries by detecting all possible
  conflicts in single preflight step: User#restoreOrBuild

vladimyr committed:
Update lockfile :goat:

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -52,112 +57,107 @@
       form: null,
-      errors: null,
-      rules
+      serverErrors: null
```
This is placeholder for error report blob so please make it clear what it holds by using more descriptive name like `errorReport` :bowtie:

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -52,112 +57,107 @@
       }).catch(err => {
         this.importing = false;
-        this.errors = true;
+        this.vErrors.add({ field: 'file', msg: 'Internal server error.' });
```
This message is not really for user's eyes. Use something more user-friendly like: `Importing users failed.`.

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -52,112 +57,107 @@
 </script>
 
 <style lang="scss" scoped>
+#userImportInput {
```
Don't style ids, add class and use it here.

vladimyr commented on client/admin/components/users/ImportDialog.vue:
```diff
@@ -1,48 +1,53 @@
-          <v-btn :disabled="disabled" color="success" type="submit">
-            Import
+          <v-btn
+            :disabled="!filename || vErrors.any() || importing"
```
Move this to computed it is bit too much logic to be inlined inside template 😉

Commit:
Cleanup :lipstick:

Commit:
Wrote a line under 80 columns :goat:

Commit:
Merge branch 'develop' into feature/import-users

Commit:
:lipstick:
